sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,263 @@
+ * @param[in] payload       pointer to the payload
+ *
+ * @return                  the checksum of the pkt in host byte order
+ * @return                  0 on error
+ */
+static uint16_t _calc_csum(ng_pktsnip_t *hdr, ng_pktsnip_t *pseudo_hdr,
+                           ng_pktsnip_t *payload)
+{
+    uint16_t csum = 0;
+    uint16_t len = (uint16_t)hdr->size;
+
+    /* process the payload */
+    while (payload && payload != hdr) {
+        csum = ng_inet_csum(csum, (uint8_t *)(payload->data), payload->size);
+        len += (uint16_t)payload->size;
+        payload = payload->next;
+    }
+    /* process applicable UDP header bytes */
+    csum = ng_inet_csum(csum, (uint8_t *)hdr->data, 6);
+
+    switch (pseudo_hdr->type) {
+#ifdef MODULE_NG_IPV6
+        case NG_NETTYPE_IPV6:
+            csum = ng_ipv6_hdr_inet_csum(csum, pseudo_hdr->data,
+                                         NG_PROTONUM_UDP, len);
+#endif
+        default:
+            return 0;
+    }
+    /* return inverted results */
+    return ~csum;
+}
+
+static void _receive(ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *udp;
+    ng_udp_hdr_t *hdr;
+    uint16_t csum;
+    uint32_t port;
+    ng_netreg_entry_t *sendto;
+
+    /* mark UDP header */
+    udp = ng_pktbuf_start_write(pkt);
+    if (udp == NULL) {
+        DEBUG("udp: unable to get write access to packet\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    pkt = udp;
+    udp = ng_pktbuf_add(pkt, pkt->data, sizeof(ng_udp_hdr_t), NG_NETTYPE_UDP);
+    if (udp == NULL) {
+        DEBUG("udp: error marking UDP header, dropping packet\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    hdr = (ng_udp_hdr_t *)udp->data;
+
+    /* validate checksum */
+    csum = _calc_csum(udp, udp->next, pkt);     /* TODO: what about extension headers? */
+    if (byteorder_ntohs(hdr->checksum) != csum) {
+        DEBUG("udp: received packet with invalid checksum, dropping it\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+    /* get port (netreg demux context) */
+    port = (uint32_t)byteorder_ntohs(hdr->dst_port);
+
+    /* send payload to receivers */
+    sendto = ng_netreg_lookup(NG_NETTYPE_UDP, port);
+    if (sendto == NULL) {
+        DEBUG("udp: unable to forward packet as no one is interested in it\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    ng_pktbuf_hold(pkt, ng_netreg_num(NG_NETTYPE_UDP, port) - 1);
+    while (sendto != NULL) {
+        ng_netapi_receive(sendto->pid, pkt);
+        sendto = ng_netreg_getnext(sendto);
+    }
+}
+
+static void _send(ng_pktsnip_t *pkt)
+{
+    ng_udp_hdr_t *hdr;
+    ng_pktsnip_t *udp_snip;
+    ng_netreg_entry_t *sendto;
+
+    /* get udp snip and hdr */
+    LL_SEARCH_SCALAR(pkt, udp_snip, type, NG_NETTYPE_UDP);
+    ng_pktbuf_start_write(udp_snip);
+    hdr = (ng_udp_hdr_t *)udp_snip->data;
+    /* fill in size field */
+    hdr->length = byteorder_htons(pkt->next->next->size + sizeof(ng_udp_hdr_t));
```

Comment:
This will fail for an empty UDP packet and for more than one snip of payload. I don't know if more than one snip of payload is legal but if it is, then:
`hdr->length = byteorder_htons(ng_pkt_len(udp_snip));`
If it is not, then the looping over `payload->next` in `_calc_csum` is  unnecessary.